File: leetcode/101-200/137-single-number-ii/single_number.py
"""LeetCode #137: Single Number II — every element appears 3x except one.

Two canonical approaches, both mirrored:
  A) ones/twos bitmask FSM: ones^=x & ~twos; twos^=x & ~ones (32-bit).
  B) per-bit count mod 3 reconstruction.
"""

MASK32 = 0xFFFFFFFF


def single_ones_twos(nums):
    ones = 0
    twos = 0
    for x in nums:
        x &= MASK32
        ones = (ones ^ x) & ~twos & MASK32
        twos = (twos ^ x) & ~ones & MASK32
    # sign-extend from 32-bit
    if ones >= 0x80000000:
        ones -= 0x100000000
    return ones


def single_bitcount(nums):
    res = 0
    for b in range(32):
        cnt = 0
        for x in nums:
            if (x >> b) & 1:
                cnt += 1
        if cnt % 3:
            res |= (1 << b)
    if res >= 0x80000000:
        res -= 0x100000000
    return res


def main():
    MOD = 1000000007
    cases = [
        [2, 2, 3, 2],
        [0, 1, 0, 1, 0, 1, 99],
        [-2, -2, 1, 1, -3, 1, -3, -3, -2, 30],
        [5],
        [7, 7, 7, 13],
    ]
    sink_a = 0
    sink_b = 0
    for nums in cases:
        a = single_ones_twos(nums)
        b = single_bitcount(nums)
        assert a == b, (nums, a, b)
        print(f"n={len(nums)}: {a}")
        sink_a = (sink_a * 1000003 + (a + 1000)) % MOD
        sink_b = (sink_b * 1000003 + (b + 1000)) % MOD
    assert sink_a == sink_b
    print(f"sink: {sink_a}")


if __name__ == "__main__":
    main()

File: leetcode/101-200/137-single-number-ii/bench/single_number.c
/* Benchmark harness for LeetCode #137 — Single Number II.
 * Mirrors single_number.kara algorithm-for-algorithm. */

#include <stdio.h>
#include <stdlib.h>

#define NP 4
#define K 30000
#define ITERS 40
#define N (3 * K + 1)

static long long mask32(void) {
    return 4294967295LL;
}

static long long sign_extend32(long long v) {
    if (v >= 2147483648LL) {
        return v - 4294967296LL;
    }
    return v;
}

static long long single_ones_twos(const long long *nums, long long n) {
    long long mask = mask32();
    long long ones = 0;
    long long twos = 0;
    for (long long i = 0; i < n; i++) {
        long long x = nums[i] & mask;
        ones = (ones ^ x) & (~twos) & mask;
        twos = (twos ^ x) & (~ones) & mask;
    }
    return sign_extend32(ones);
}

static long long single_bitcount(const long long *nums, long long n) {
    long long res = 0;
    for (long long b = 0; b < 32; b++) {
        long long cnt = 0;
        for (long long i = 0; i < n; i++) {
            if (((nums[i] >> b) & 1LL) == 1LL) {
                cnt++;
            }
        }
        if ((cnt % 3) != 0) {
            res |= 1LL << b;
        }
    }
    return sign_extend32(res);
}

int main(void) {
    static long long arrays[NP][N];

    for (long long j = 0; j < NP; j++) {
        static long long vals[K];
        long long x = j + 1;
        for (long long t = 0; t < K; t++) {
            x = (x * 1103515245LL + 12345LL) % 2147483648LL;
            long long hi = x / 65536LL;
            x = (x * 1103515245LL + 12345LL) % 2147483648LL;
            vals[t] = (hi * 32768LL + x / 65536LL) % 100000LL;
        }
        long long e = 0;
        for (long long pass = 0; pass < 3; pass++) {
            for (long long q = 0; q < K; q++) {
                arrays[j][e++] = vals[q];
            }
        }
        arrays[j][e++] = 999983LL + j;
    }

    long long sink = 0;
    for (long long it = 0; it < ITERS; it++) {
        long long idx = (it * 3) % NP;
        long long a = single_ones_twos(arrays[idx], N);
        long long b = single_bitcount(arrays[idx], N);
        if (a != b) {
            sink += 1000000000LL;
        }
        sink = (sink + a + b) % 1000000007LL;
    }
    printf("%lld\n", sink);
    return 0;
}
